packages/qwik-city/runtime/src/single-page-navigation-component.tsx

```diff
@@ -0,0 +1,82 @@
+import {
+  component$,
+  type PropFunction,
+  useContext,
+  useOnDocument,
+  useOnWindow,
+  useSignal,
+  _IMMUTABLE,
+  _jsxBranch,
+  event$,
+} from '@builder.io/qwik';
+import { dispatchNavigationDOMSettledEvent, dispatchNavigationStatusEvent } from './event';
+import { RouteInternalContext } from './contexts';
+import type { RestoreScroll, NavigationStatusEvent } from './types';
+import { useLocation } from './use-functions';
+import { toPath } from './utils';
+
+export interface SinglePageNavigationProps {
+  restoreScroll$: PropFunction<RestoreScroll>;
+}
+
+export const SinglePageNavigation = component$<SinglePageNavigationProps>(({ restoreScroll$ }) => {
+  const routeInternal = useContext(RouteInternalContext);
+  const loc = useLocation();
+  // a rendering marker for MutationObserver to detect when the DOM is settled after nagivation
+  const marker = useSignal<HTMLScriptElement>();
+  // a promise that resolves when the DOM is settled after each navigation
+  const settled = useSignal<Promise<void>>(() => Promise.resolve());
+  // for temporarily disabling native scroll restoration in some browsers
+  const scrollRestorationSettingBackup = useSignal<ScrollRestoration>('auto');
+
+  useOnWindow(
+    'popstate',
+    event$(() => {
+      const toUrl = new URL(location.href);
+      dispatchNavigationStatusEvent('popstate', loc.url, toUrl, 'pending');
+      routeInternal.value = { type: 'popstate', dest: toUrl };
+    })
+  );
+
+  useOnDocument(
+    'qnav',
+    event$(async (e: Event) => {
+      const detail = (e as NavigationStatusEvent).detail;
+      switch (detail.status) {
+        case 'pending': {
+          // set up a single-use MutationObserver asynchrously and
+          // observe the data-key attribute of the marker for DOM update detection
+          settled.value = new Promise((resolve, reject) => {
```

Comment:
i would like to understand this. why do we need a mutation observer?

Reply:
I think both Point 3 in the PR and the comment above this line provide an explanation. Could you clarify which part you don't understand?

Reply:
This is now moved to `router-outlet-component.ts`